In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


pd.set_option("display.max_columns", None)

print("Libraries imported successfully!")

Libraries imported successfully!


# =====================================================
# Project: Salary Data Cleaning & Analysis
# Author: Abahoor Lion
# Goal:
# Combine multiple salary datasets, clean the data,
# merge into one master dataset, analyze it using SQL,
# and build an interactive Power BI dashboard.
# =====================================================

# Project Notes

## Dataset_2

salary_currency: The currency in which the salary is paid, such as USD, EUR, etc. This is important for currency conversion and understanding the actual value of the salary in a global context.

salary: The annual gross salary of the role in the local currency. This raw salary figure is key for direct regional salary comparisons.

salary_in_usd: The annual gross salary converted to United States Dollars (USD). This uniform currency conversion aids in global salary comparisons and analyses.

## Dataset_3

experience_level: The experience level in the job during the year (EN: Entry-level/Junior, MI: Mid-level/Intermediate, SE: Senior-level/Expert, EX: Executive-level/Director).

employment_type: The type of employment for the role (PT: Part-time, FT: Full-time, CT: Contract, FL: Freelance).

salary: The total gross salary amount paid.
salary_currency: The currency of the salary paid as an ISO 4217 currency code.
salary_in_usd: The salary in USD (FX rate divided by avg. USD rate of respective year).

remote_ratio: The overall amount of work done remotely (0: No remote work, 50: Partially remote/hybrid, 100: Fully remote).

company_size: The average number of people that worked for the company during the year (S: less than 50 employees, M: 50 to 250 employees, L: more than 250 employees).



In [2]:
df_1=pd.read_csv("../data/raw/DataAnalyst_set1_raw.csv")
df_1.head(10)


,Unnamed: 0,Job Title,Salary Estimate,Job Description,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply
0,0,"Data Analyst, Center on Immigration and Justic...",$37K-$66K (Glassdoor est.),Are you eager to roll up your sleeves and harn...,3.2,Vera Institute of Justice\n3.2,"New York, NY","New York, NY",201 to 500 employees,1961,Nonprofit Organization,Social Assistance,Non-Profit,$100 to $500 million (USD),-1,True
1,1,Quality Data Analyst,$37K-$66K (Glassdoor est.),Overview\n\nProvides analytical and technical ...,3.8,Visiting Nurse Service of New York\n3.8,"New York, NY","New York, NY",10000+ employees,1893,Nonprofit Organization,Health Care Services & Hospitals,Health Care,$2 to $5 billion (USD),-1,-1
2,2,"Senior Data Analyst, Insights & Analytics Team...",$37K-$66K (Glassdoor est.),We’re looking for a Senior Data Analyst who ha...,3.4,Squarespace\n3.4,"New York, NY","New York, NY",1001 to 5000 employees,2003,Company - Private,Internet,Information Technology,Unknown / Non-Applicable,GoDaddy,-1
3,3,Data Analyst,$37K-$66K (Glassdoor est.),Requisition NumberRR-0001939\nRemote:Yes\nWe c...,4.1,Celerity\n4.1,"New York, NY","McLean, VA",201 to 500 employees,2002,Subsidiary or Business Segment,IT Services,Information Technology,$50 to $100 million (USD),-1,-1
4,4,Reporting Data Analyst,$37K-$66K (Glassdoor est.),ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...,3.9,FanDuel\n3.9,"New York, NY","New York, NY",501 to 1000 employees,2009,Company - Private,Sports & Recreation,"Arts, Entertainment & Recreation",$100 to $500 million (USD),DraftKings,True
5,5,Data Analyst,$37K-$66K (Glassdoor est.),About Cubist\nCubist Systematic Strategies is ...,3.9,Point72\n3.9,"New York, NY","Stamford, CT",1001 to 5000 employees,2014,Company - Private,Investment Banking & Asset Management,Finance,Unknown / Non-Applicable,-1,-1
6,6,Business/Data Analyst (FP&A),$37K-$66K (Glassdoor est.),Two Sigma is a different kind of investment ma...,4.4,Two Sigma\n4.4,"New York, NY","New York, NY",1001 to 5000 employees,2001,Company - Private,Investment Banking & Asset Management,Finance,Unknown / Non-Applicable,-1,-1
7,7,Data Science Analyst,$37K-$66K (Glassdoor est.),Data Science Analyst\n\nJob Details\nLevel\nEx...,3.7,GNY Insurance Companies\n3.7,"New York, NY","New York, NY",201 to 500 employees,1914,Company - Private,Insurance Carriers,Insurance,$100 to $500 million (USD),"Travelers, Chubb, Crum & Forster",True
8,8,Data Analyst,$37K-$66K (Glassdoor est.),The Data Analyst is an integral member of the ...,4.0,DMGT\n4.0,"New York, NY","London, United Kingdom",5001 to 10000 employees,1896,Company - Public,Venture Capital & Private Equity,Finance,$1 to $2 billion (USD),"Thomson Reuters, Hearst, Pearson",-1
9,9,"Data Analyst, Merchant Health",$37K-$66K (Glassdoor est.),About Us\n\nRiskified is the AI platform power...,4.4,Riskified\n4.4,"New York, NY","New York, NY",501 to 1000 employees,2013,Company - Private,Research & Development,Business Services,Unknown / Non-Applicable,"Signifyd, Forter",-1


In [8]:
df_1.isnull().sum()

Unnamed: 0           0
Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         1
Location             0
Headquarters         0
Size                 0
Founded              0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
Competitors          0
Easy Apply           0
dtype: int64

In [5]:
df_2=pd.read_csv("../data/raw/jobs_in_data_set2_raw.csv")
df_2.head(10)

,work_year,job_title,job_category,salary_currency,salary,salary_in_usd,employee_residence,experience_level,employment_type,work_setting,company_location,company_size
0,2023,Data DevOps Engineer,Data Engineering,EUR,88000,95012,Germany,Mid-level,Full-time,Hybrid,Germany,L
1,2023,Data Architect,Data Architecture and Modeling,USD,186000,186000,United States,Senior,Full-time,In-person,United States,M
2,2023,Data Architect,Data Architecture and Modeling,USD,81800,81800,United States,Senior,Full-time,In-person,United States,M
3,2023,Data Scientist,Data Science and Research,USD,212000,212000,United States,Senior,Full-time,In-person,United States,M
4,2023,Data Scientist,Data Science and Research,USD,93300,93300,United States,Senior,Full-time,In-person,United States,M
5,2023,Data Scientist,Data Science and Research,USD,130000,130000,United States,Senior,Full-time,Remote,United States,M
6,2023,Data Scientist,Data Science and Research,USD,100000,100000,United States,Senior,Full-time,Remote,United States,M
7,2023,Machine Learning Researcher,Machine Learning and AI,USD,224400,224400,United States,Mid-level,Full-time,In-person,United States,M
8,2023,Machine Learning Researcher,Machine Learning and AI,USD,138700,138700,United States,Mid-level,Full-time,In-person,United States,M
9,2023,Data Engineer,Data Engineering,USD,210000,210000,United States,Executive,Full-time,Remote,United States,M


In [7]:
df_2.isnull().sum()

work_year             0
job_title             0
job_category          0
salary_currency       0
salary                0
salary_in_usd         0
employee_residence    0
experience_level      0
employment_type       0
work_setting          0
company_location      0
company_size          0
dtype: int64

In [2]:
df_3=pd.read_csv("../data/raw/global_ai_ml_data_salaries_set3_raw.csv")
df_3.head(10)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2024,SE,FT,Data Architect,170955,USD,170955,US,0,US,M
1,2024,SE,FT,Data Architect,130270,USD,130270,US,0,US,M
2,2024,MI,FT,Insight Analyst,92144,USD,92144,US,100,US,M
3,2024,MI,FT,Insight Analyst,57616,USD,57616,US,100,US,M
4,2024,SE,FT,Lead Engineer,201600,USD,201600,US,100,US,M
5,2024,SE,FT,Lead Engineer,181440,USD,181440,US,100,US,M
6,2024,SE,FT,Machine Learning Engineer,296300,USD,296300,US,0,US,M
7,2024,SE,FT,Machine Learning Engineer,166600,USD,166600,US,0,US,M
8,2024,SE,FT,Machine Learning Engineer,264200,USD,264200,US,0,US,M
9,2024,SE,FT,Machine Learning Engineer,143100,USD,143100,US,0,US,M


In [16]:
df_3.isnull().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

In [23]:
print(df_1.shape)
print("---------------")
print(df_1.info())

(2253, 16)
---------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2253 entries, 0 to 2252
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         2253 non-null   int64  
 1   Job Title          2253 non-null   object 
 2   Salary Estimate    2253 non-null   object 
 3   Job Description    2253 non-null   object 
 4   Rating             2253 non-null   float64
 5   Company Name       2252 non-null   object 
 6   Location           2253 non-null   object 
 7   Headquarters       2253 non-null   object 
 8   Size               2253 non-null   object 
 9   Founded            2253 non-null   int64  
 10  Type of ownership  2253 non-null   object 
 11  Industry           2253 non-null   object 
 12  Sector             2253 non-null   object 
 13  Revenue            2253 non-null   object 
 14  Competitors        2253 non-null   object 
 15  Easy Apply         2253 non-null   object 
dt

In [25]:
print(df_2.shape)
print("--------")
print(df_2.info())

(9355, 12)
--------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9355 entries, 0 to 9354
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           9355 non-null   int64 
 1   job_title           9355 non-null   object
 2   job_category        9355 non-null   object
 3   salary_currency     9355 non-null   object
 4   salary              9355 non-null   int64 
 5   salary_in_usd       9355 non-null   int64 
 6   employee_residence  9355 non-null   object
 7   experience_level    9355 non-null   object
 8   employment_type     9355 non-null   object
 9   work_setting        9355 non-null   object
 10  company_location    9355 non-null   object
 11  company_size        9355 non-null   object
dtypes: int64(3), object(9)
memory usage: 877.2+ KB
None


In [26]:
print(df_3.shape)
print("-----------")
print(df_3.info())
print("--------------")
print(df_3.describe())

(40332, 11)
-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40332 entries, 0 to 40331
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           40332 non-null  int64 
 1   experience_level    40332 non-null  object
 2   employment_type     40332 non-null  object
 3   job_title           40332 non-null  object
 4   salary              40332 non-null  int64 
 5   salary_currency     40332 non-null  object
 6   salary_in_usd       40332 non-null  int64 
 7   employee_residence  40332 non-null  object
 8   remote_ratio        40332 non-null  int64 
 9   company_location    40332 non-null  object
 10  company_size        40332 non-null  object
dtypes: int64(4), object(7)
memory usage: 3.4+ MB
None
--------------
          work_year        salary  salary_in_usd  remote_ratio
count  40332.000000  4.033200e+04   40332.000000  40332.000000
mean    2023.682832  1.667671e+05  160758.51963

In [27]:
print(df_1.duplicated().sum())
print(df_2.duplicated().sum())
print(df_3.duplicated().sum())

0
4014
19925


In [11]:
df_1.columns


Index(['Job Title', 'Salary Estimate', 'Rating', 'Company Name', 'Location',
       'Headquarters', 'Size', 'Founded', 'Type of ownership', 'Industry',
       'Sector', 'Revenue', 'Competitors', 'Easy Apply'],
      dtype='object')

In [29]:
df_2.columns


Index(['work_year', 'job_title', 'job_category', 'salary_currency', 'salary',
       'salary_in_usd', 'employee_residence', 'experience_level',
       'employment_type', 'work_setting', 'company_location', 'company_size'],
      dtype='object')

In [30]:
df_3.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

In [6]:
df_1.drop(columns=["Unnamed: 0"], inplace=True)

In [ ]:
df_1.drop(columns=["Job Description"],inplace=True)

In [12]:
df_1["Easy Apply"].value_counts(dropna=False)

Easy Apply
-1      2173
True      80
Name: count, dtype: int64

In [13]:
df_1["Competitors"].value_counts(dropna=False)

Competitors
-1                                                           1732
Adecco, Manpower                                               14
Robert Half, Insight Global                                    14
Google, Microsoft, Samsung Electronics                         10
Artech Information Systems, Mindlance, Tech Mahindra           10
                                                             ... 
Cognizant Technology Solutions, Infosys, HCL Technologies       1
Equinix, Digital Realty, CyrusOne                               1
WeDo technologies, Intec Telecom, Amdocs                        1
Thomson Reuters, Hearst, Pearson                                1
Travelers, Chubb, Crum & Forster                                1
Name: count, Length: 291, dtype: int64

In [20]:
df_1["Easy Apply"].unique

<bound method Series.unique of 0       True
1         -1
2         -1
3         -1
4       True
        ... 
2248      -1
2249      -1
2250      -1
2251      -1
2252      -1
Name: Easy Apply, Length: 2253, dtype: object>

In [22]:
#Every val is stored as str
#Look at the typ of the value beffor changing it
df_1["Easy Apply"].apply(type).value_counts()

Easy Apply
<class 'str'>    2253
Name: count, dtype: int64

In [23]:
df_1["Easy Apply"]=df_1["Easy Apply"].replace({"True": 1 , "-1": 0})

C:\Users\abaho\AppData\Local\Temp\ipykernel_6392\1468928158.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_1["Easy Apply"]=df_1["Easy Apply"].replace({"True": 1 , "-1": 0})


In [24]:
df_1["Easy Apply"].value_counts(dropna=False)

Easy Apply
0    2173
1      80
Name: count, dtype: int64

In [26]:
df_1["Easy Apply"].dtype

dtype('int64')

### Data Cleaning Step 1: Easy Apply

**Issue:**
The `Easy Apply` column contained `True` and `-1`, which are not ideal for SQL or Power BI.

**Solution:**
Converted:
- `True` → `1`
- `-1` → `0`

**Reason:**
A binary (0/1) column is easier to analyze, filter, and aggregate in SQL and Power BI.

In [14]:
df_1["Salary Estimate"].head(10)

0    $37K-$66K (Glassdoor est.)
1    $37K-$66K (Glassdoor est.)
2    $37K-$66K (Glassdoor est.)
3    $37K-$66K (Glassdoor est.)
4    $37K-$66K (Glassdoor est.)
5    $37K-$66K (Glassdoor est.)
6    $37K-$66K (Glassdoor est.)
7    $37K-$66K (Glassdoor est.)
8    $37K-$66K (Glassdoor est.)
9    $37K-$66K (Glassdoor est.)
Name: Salary Estimate, dtype: object

In [ ]:
df_1["Salary Estimate"].tail(10)
#The values are sorted

2243    $78K-$104K (Glassdoor est.)
2244    $78K-$104K (Glassdoor est.)
2245    $78K-$104K (Glassdoor est.)
2246    $78K-$104K (Glassdoor est.)
2247    $78K-$104K (Glassdoor est.)
2248    $78K-$104K (Glassdoor est.)
2249    $78K-$104K (Glassdoor est.)
2250    $78K-$104K (Glassdoor est.)
2251    $78K-$104K (Glassdoor est.)
2252    $78K-$104K (Glassdoor est.)
Name: Salary Estimate, dtype: object

In [ ]:
# Use value_counts to see the most commend values.
# Use head() function to see the most common value and tail() for the least common value 
df_1["Salary Estimate"].value_counts().head(10)

Salary Estimate
$41K-$78K (Glassdoor est.)     57
$42K-$76K (Glassdoor est.)     57
$50K-$86K (Glassdoor est.)     41
$35K-$67K (Glassdoor est.)     33
$60K-$124K (Glassdoor est.)    31
$58K-$93K (Glassdoor est.)     31
$43K-$76K (Glassdoor est.)     31
$46K-$87K (Glassdoor est.)     30
$51K-$87K (Glassdoor est.)     30
$37K-$66K (Glassdoor est.)     30
Name: count, dtype: int64

In [12]:
df_1["Salary Estimate"].value_counts().tail(10)

Salary Estimate
$40K-$74K (Glassdoor est.)    9
$30K-$54K (Glassdoor est.)    8
$42K-$63K (Glassdoor est.)    4
$31K-$59K (Glassdoor est.)    4
$32K-$56K (Glassdoor est.)    4
$47K-$81K (Glassdoor est.)    3
$36K-$67K (Glassdoor est.)    3
$43K-$77K (Glassdoor est.)    3
$57K-$70K (Glassdoor est.)    2
-1                            1
Name: count, dtype: int64

In [15]:
df_1[df_1["Salary Estimate"] == "-1"]

,Unnamed: 0,Job Title,Salary Estimate,Job Description,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply
2149,2149,Configuration/Data Management Analyst III,-1,Job Description\nPosition Title: Configuration...,4.4,Protingent\n4.4,"Kent, WA","Bellevue, WA",51 to 200 employees,2001,Company - Private,Staffing & Outsourcing,Business Services,$25 to $50 million (USD),-1,-1


In [17]:
df_1["Salary Estimate"].sample(20,random_state=42)

1785     $28K-$52K (Glassdoor est.)
1263    $76K-$122K (Glassdoor est.)
2126    $55K-$101K (Glassdoor est.)
599     $57K-$103K (Glassdoor est.)
464      $43K-$69K (Glassdoor est.)
1264    $76K-$122K (Glassdoor est.)
1745     $50K-$86K (Glassdoor est.)
1960    $93K-$159K (Glassdoor est.)
1566     $51K-$93K (Glassdoor est.)
1270    $76K-$122K (Glassdoor est.)
134      $59K-$85K (Glassdoor est.)
613      $35K-$45K (Glassdoor est.)
188     $60K-$110K (Glassdoor est.)
1317     $24K-$48K (Glassdoor est.)
1928    $99K-$178K (Glassdoor est.)
1985    $82K-$116K (Glassdoor est.)
2094     $34K-$61K (Glassdoor est.)
445      $26K-$47K (Glassdoor est.)
410      $48K-$96K (Glassdoor est.)
1061     $45K-$78K (Glassdoor est.)
Name: Salary Estimate, dtype: object

In [35]:
df_1["salary_clean"] = df_1["Salary Estimate"]

In [62]:
df_1["Salary Estimate"] = df_1["Salary Estimate"].str.replace("(Glassdoor est.)","")
print(df_1["Salary Estimate"] )

0        37-66 
1        37-66 
2        37-66 
3        37-66 
4        37-66 
         ...   
2248    78-104 
2249    78-104 
2250    78-104 
2251    78-104 
2252    78-104 
Name: Salary Estimate, Length: 2253, dtype: object


In [63]:
df_1["Salary Estimate"].head()

0    37-66 
1    37-66 
2    37-66 
3    37-66 
4    37-66 
Name: Salary Estimate, dtype: object

In [64]:
df_1["Salary Estimate"]=df_1["Salary Estimate"].str.replace("$","")

In [65]:
df_1["Salary Estimate"] =df_1["Salary Estimate"] .str.replace("K","")
print(df_1["Salary Estimate"] )

0        37-66 
1        37-66 
2        37-66 
3        37-66 
4        37-66 
         ...   
2248    78-104 
2249    78-104 
2250    78-104 
2251    78-104 
2252    78-104 
Name: Salary Estimate, Length: 2253, dtype: object


In [90]:
# Creat two colum out of one colum
df_1[["salary_min","salary_max"]] = df_1["Salary Estimate"].str.split("-",expand=True) 

In [67]:
df_1["salary_min"]

0       37
1       37
2       37
3       37
4       37
        ..
2248    78
2249    78
2250    78
2251    78
2252    78
Name: salary_min, Length: 2253, dtype: object

In [68]:
df_1[["Salary Estimate", "salary_min", "salary_max"]].head()

,Salary Estimate,salary_min,salary_max
0,37-66,37,66
1,37-66,37,66
2,37-66,37,66
3,37-66,37,66
4,37-66,37,66


In [69]:
df_1["salary_min"].apply(type)

0       <class 'str'>
1       <class 'str'>
2       <class 'str'>
3       <class 'str'>
4       <class 'str'>
            ...      
2248    <class 'str'>
2249    <class 'str'>
2250    <class 'str'>
2251    <class 'str'>
2252    <class 'str'>
Name: salary_min, Length: 2253, dtype: object

In [70]:
df_1["salary_min"].astype(int)
df_1["salary_max"].astype(int)

ValueError: invalid literal for int() with base 10: ''

In [71]:
df_1[df_1["salary_min"] == ""]

,Unnamed: 0,Job Title,Salary Estimate,Job Description,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply,salary_min,salary_max
2149,2149,Configuration/Data Management Analyst III,-1,Job Description\nPosition Title: Configuration...,4.4,Protingent\n4.4,"Kent, WA","Bellevue, WA",51 to 200 employees,2001,Company - Private,Staffing & Outsourcing,Business Services,$25 to $50 million (USD),-1,-1,,1


In [29]:
df_1[df_1["salary_max"] == ""]

,Unnamed: 0,Job Title,Salary Estimate,Job Description,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply,salary_min,salary_max


In [72]:
bad_row = df_1[df_1["Salary Estimate"] == "-1"].index

In [73]:
df_1=df_1.drop(bad_row)

In [74]:
df_1[df_1["Salary Estimate"] == "-1"]

,Unnamed: 0,Job Title,Salary Estimate,Job Description,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply,salary_min,salary_max


In [91]:
df_1["salary_min"]=df_1["salary_min"].astype(int)
df_1["salary_max"]=df_1["salary_max"].astype(int)

In [92]:
df_1["salary_min"].apply(type)

0       <class 'int'>
1       <class 'int'>
2       <class 'int'>
3       <class 'int'>
4       <class 'int'>
            ...      
2248    <class 'int'>
2249    <class 'int'>
2250    <class 'int'>
2251    <class 'int'>
2252    <class 'int'>
Name: salary_min, Length: 2252, dtype: object

In [93]:
df_1["salary_max"].apply(type)

0       <class 'int'>
1       <class 'int'>
2       <class 'int'>
3       <class 'int'>
4       <class 'int'>
            ...      
2248    <class 'int'>
2249    <class 'int'>
2250    <class 'int'>
2251    <class 'int'>
2252    <class 'int'>
Name: salary_max, Length: 2252, dtype: object

In [94]:
df_1["salary_avg"]=(df_1["salary_min"]+df_1["salary_max"])/2

In [95]:
df_1[["salary_min", "salary_max", "salary_avg"]].head()

,salary_min,salary_max,salary_avg
0,37,66,51.5
1,37,66,51.5
2,37,66,51.5
3,37,66,51.5
4,37,66,51.5


In [96]:
df_1.columns

Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'Job Description',
       'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply', 'salary_min', 'salary_max', 'salary_avg'],
      dtype='object')

In [97]:
col1=df_1.pop("salary_min")
col2=df_1.pop("salary_max")
col3=df_1.pop("salary_avg")
df_1.insert(3,"salary_min",col1)
df_1.insert(4,"salary_max",col2)
df_1.insert(5,"salary_avg",col3)

In [98]:
df_1.columns

Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'salary_min',
       'salary_max', 'salary_avg', 'Job Description', 'Rating', 'Company Name',
       'Location', 'Headquarters', 'Size', 'Founded', 'Type of ownership',
       'Industry', 'Sector', 'Revenue', 'Competitors', 'Easy Apply'],
      dtype='object')

In [3]:
df_test = pd.read_csv("../data/cleaned/data_analyst_set1_cleaned.csv")

df_test.head()

,Job Title,Salary Estimate,salary_min,salary_max,salary_avg,Rating,Company Name,Location,Headquarters,Size,Founded,Type of ownership,Industry,Sector,Revenue,Competitors,Easy Apply
0,"Data Analyst, Center on Immigration and Justic...",37-66,37,66,51.5,3.2,Vera Institute of Justice\r\n3.2,"New York, NY","New York, NY",201 to 500 employees,1961,Nonprofit Organization,Social Assistance,Non-Profit,$100 to $500 million (USD),-1,1
1,Quality Data Analyst,37-66,37,66,51.5,3.8,Visiting Nurse Service of New York\r\n3.8,"New York, NY","New York, NY",10000+ employees,1893,Nonprofit Organization,Health Care Services & Hospitals,Health Care,$2 to $5 billion (USD),-1,0
2,"Senior Data Analyst, Insights & Analytics Team...",37-66,37,66,51.5,3.4,Squarespace\r\n3.4,"New York, NY","New York, NY",1001 to 5000 employees,2003,Company - Private,Internet,Information Technology,Unknown / Non-Applicable,GoDaddy,0
3,Data Analyst,37-66,37,66,51.5,4.1,Celerity\r\n4.1,"New York, NY","McLean, VA",201 to 500 employees,2002,Subsidiary or Business Segment,IT Services,Information Technology,$50 to $100 million (USD),-1,0
4,Reporting Data Analyst,37-66,37,66,51.5,3.9,FanDuel\r\n3.9,"New York, NY","New York, NY",501 to 1000 employees,2009,Company - Private,Sports & Recreation,"Arts, Entertainment & Recreation",$100 to $500 million (USD),DraftKings,1


In [4]:
df_test["Size"].value_counts()

Size
51 to 200 employees        420
10000+ employees           375
1001 to 5000 employees     348
1 to 50 employees          347
201 to 500 employees       249
501 to 1000 employees      211
-1                         163
5001 to 10000 employees     97
Unknown                     42
Name: count, dtype: int64

Clean Dataset 1

- Removed unnecessary columns
- Converted Easy Apply to binary
- Cleaned Salary Estimate column
- Created salary_min, salary_max, salary_avg
- Removed invalid salary record
- Reordered columns
- Exported cleaned dataset

In [6]:
df_2.head(10)

,work_year,job_title,job_category,salary_currency,salary,salary_in_usd,employee_residence,experience_level,employment_type,work_setting,company_location,company_size
0,2023,Data DevOps Engineer,Data Engineering,EUR,88000,95012,Germany,Mid-level,Full-time,Hybrid,Germany,L
1,2023,Data Architect,Data Architecture and Modeling,USD,186000,186000,United States,Senior,Full-time,In-person,United States,M
2,2023,Data Architect,Data Architecture and Modeling,USD,81800,81800,United States,Senior,Full-time,In-person,United States,M
3,2023,Data Scientist,Data Science and Research,USD,212000,212000,United States,Senior,Full-time,In-person,United States,M
4,2023,Data Scientist,Data Science and Research,USD,93300,93300,United States,Senior,Full-time,In-person,United States,M
5,2023,Data Scientist,Data Science and Research,USD,130000,130000,United States,Senior,Full-time,Remote,United States,M
6,2023,Data Scientist,Data Science and Research,USD,100000,100000,United States,Senior,Full-time,Remote,United States,M
7,2023,Machine Learning Researcher,Machine Learning and AI,USD,224400,224400,United States,Mid-level,Full-time,In-person,United States,M
8,2023,Machine Learning Researcher,Machine Learning and AI,USD,138700,138700,United States,Mid-level,Full-time,In-person,United States,M
9,2023,Data Engineer,Data Engineering,USD,210000,210000,United States,Executive,Full-time,Remote,United States,M


In [7]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9355 entries, 0 to 9354
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           9355 non-null   int64 
 1   job_title           9355 non-null   object
 2   job_category        9355 non-null   object
 3   salary_currency     9355 non-null   object
 4   salary              9355 non-null   int64 
 5   salary_in_usd       9355 non-null   int64 
 6   employee_residence  9355 non-null   object
 7   experience_level    9355 non-null   object
 8   employment_type     9355 non-null   object
 9   work_setting        9355 non-null   object
 10  company_location    9355 non-null   object
 11  company_size        9355 non-null   object
dtypes: int64(3), object(9)
memory usage: 877.2+ KB


In [8]:
df_2.columns

Index(['work_year', 'job_title', 'job_category', 'salary_currency', 'salary',
       'salary_in_usd', 'employee_residence', 'experience_level',
       'employment_type', 'work_setting', 'company_location', 'company_size'],
      dtype='object')

In [10]:
print(df_2.shape)
print(df_2.duplicated().sum())


(9355, 12)
4014


In [11]:
df_2["company_size"].value_counts()

company_size
M    8448
L     748
S     159
Name: count, dtype: int64

In [13]:
df_2["company_size"] = df_2["company_size"].replace("S","Small")
df_2["company_size"] = df_2["company_size"].replace("M","Medium")
df_2["company_size"] = df_2["company_size"].replace("L","Large")

In [14]:
df_2["company_size"].value_counts()

company_size
Medium    8448
Large      748
Small      159
Name: count, dtype: int64

In [15]:
duplicate_rows = df_2[df_2.duplicated(keep=False)]

duplicate_rows.sort_values(by=["job_title", "salary_in_usd", "employee_residence"]).head(20)

,work_year,job_title,job_category,salary_currency,salary,salary_in_usd,employee_residence,experience_level,employment_type,work_setting,company_location,company_size
3157,2023,AI Architect,Machine Learning and AI,USD,204000,204000,United States,Senior,Full-time,In-person,United States,Medium
3968,2023,AI Architect,Machine Learning and AI,USD,204000,204000,United States,Senior,Full-time,In-person,United States,Medium
4308,2023,AI Architect,Machine Learning and AI,USD,204000,204000,United States,Senior,Full-time,In-person,United States,Medium
4506,2023,AI Architect,Machine Learning and AI,USD,204000,204000,United States,Senior,Full-time,In-person,United States,Medium
3156,2023,AI Architect,Machine Learning and AI,USD,330000,330000,United States,Senior,Full-time,In-person,United States,Medium
3967,2023,AI Architect,Machine Learning and AI,USD,330000,330000,United States,Senior,Full-time,In-person,United States,Medium
4307,2023,AI Architect,Machine Learning and AI,USD,330000,330000,United States,Senior,Full-time,In-person,United States,Medium
4505,2023,AI Architect,Machine Learning and AI,USD,330000,330000,United States,Senior,Full-time,In-person,United States,Medium
1887,2023,AI Engineer,Machine Learning and AI,USD,83900,83900,United States,Senior,Full-time,In-person,United States,Medium
1923,2023,AI Engineer,Machine Learning and AI,USD,83900,83900,United States,Senior,Full-time,In-person,United States,Medium


In [16]:
df_2.drop_duplicates().shape

(5341, 12)

In [17]:
df_2.nunique()

work_year                4
job_title              125
job_category            10
salary_currency         11
salary                1507
salary_in_usd         1786
employee_residence      83
experience_level         4
employment_type          4
work_setting             3
company_location        70
company_size             3
dtype: int64

## Dataset 2 Cleaning Summary

### Changes Made

- Converted `company_size` values into descriptive labels:
  - `S` → `Small`
  - `M` → `Medium`
  - `L` → `Large`

### Data Quality Checks

- Checked for missing values.
- Checked for duplicate records.
- Found 4,014 duplicate rows.

### Duplicate Decision

The duplicate rows were **not removed**.

Reason:
This dataset does not contain a unique employee ID or company ID. Because of that, multiple employees may legitimately share the same values (job title, salary, country, experience level, etc.). Removing duplicates could accidentally delete valid records.

### Result

The dataset required very little cleaning compared to Dataset 1 and was already well structured.

In [5]:
df_3.head(10)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2024,SE,FT,Data Architect,170955,USD,170955,US,0,US,M
1,2024,SE,FT,Data Architect,130270,USD,130270,US,0,US,M
2,2024,MI,FT,Insight Analyst,92144,USD,92144,US,100,US,M
3,2024,MI,FT,Insight Analyst,57616,USD,57616,US,100,US,M
4,2024,SE,FT,Lead Engineer,201600,USD,201600,US,100,US,M
5,2024,SE,FT,Lead Engineer,181440,USD,181440,US,100,US,M
6,2024,SE,FT,Machine Learning Engineer,296300,USD,296300,US,0,US,M
7,2024,SE,FT,Machine Learning Engineer,166600,USD,166600,US,0,US,M
8,2024,SE,FT,Machine Learning Engineer,264200,USD,264200,US,0,US,M
9,2024,SE,FT,Machine Learning Engineer,143100,USD,143100,US,0,US,M


In [6]:
print(df_3["experience_level"].value_counts())
print("--------------------")
print(df_3["employment_type"].value_counts())
print("-------------------------")
print(df_3["company_size"].value_counts())

experience_level
SE    24236
MI    11731
EN     3477
EX      888
Name: count, dtype: int64
--------------------
employment_type
FT    40203
PT       60
CT       55
FL       14
Name: count, dtype: int64
-------------------------
company_size
M    38611
L     1521
S      200
Name: count, dtype: int64


In [7]:
df_3["experience_level"] = df_3["experience_level"].replace("SE","Senior level")
df_3["experience_level"] = df_3["experience_level"].replace("MI","Mid level")
df_3["experience_level"] = df_3["experience_level"].replace("EN","Entry level")
df_3["experience_level"] = df_3["experience_level"].replace("EX","Executive level")

In [8]:
df_3["employment_type"] = df_3["employment_type"].replace("FT","Full time")
df_3["employment_type"] = df_3["employment_type"].replace("PT","Part-time")
df_3["employment_type"] = df_3["employment_type"].replace("CT","Contract")
df_3["employment_type"] = df_3["employment_type"].replace("FL","Freelance")

In [9]:
df_3["company_size"] = df_3["company_size"].replace("L","Large")
df_3["company_size"] = df_3["company_size"].replace("M","Medium")
df_3["company_size"] = df_3["company_size"].replace("S","Small")

In [10]:
df_3.head(10)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2024,Senior level,Full time,Data Architect,170955,USD,170955,US,0,US,Medium
1,2024,Senior level,Full time,Data Architect,130270,USD,130270,US,0,US,Medium
2,2024,Mid level,Full time,Insight Analyst,92144,USD,92144,US,100,US,Medium
3,2024,Mid level,Full time,Insight Analyst,57616,USD,57616,US,100,US,Medium
4,2024,Senior level,Full time,Lead Engineer,201600,USD,201600,US,100,US,Medium
5,2024,Senior level,Full time,Lead Engineer,181440,USD,181440,US,100,US,Medium
6,2024,Senior level,Full time,Machine Learning Engineer,296300,USD,296300,US,0,US,Medium
7,2024,Senior level,Full time,Machine Learning Engineer,166600,USD,166600,US,0,US,Medium
8,2024,Senior level,Full time,Machine Learning Engineer,264200,USD,264200,US,0,US,Medium
9,2024,Senior level,Full time,Machine Learning Engineer,143100,USD,143100,US,0,US,Medium


# Dataset 3 Cleaning Summary

## Changes Made

### experience_level

Converted abbreviated experience levels into descriptive values.

- EN → Entry level
- MI → Mid level
- SE → Senior level
- EX → Executive level

---

### employment_type

Converted abbreviated employment types into descriptive values.

- FT → Full time
- PT → Part-time
- CT → Contract
- FL → Freelance

---

### company_size

Converted abbreviated company size values into descriptive labels.

- S → Small
- M → Medium
- L → Large

---

### remote_ratio

The `remote_ratio` column was left unchanged.

Reason:

The values already represent meaningful numeric categories.

- 0 = On-site
- 50 = Hybrid
- 100 = Fully remote

Keeping the column numeric makes it easier to analyze in SQL and Power BI while preserving the ordinal relationship between work settings.

---

## Data Quality Checks

- Checked missing values.
- Checked duplicate rows.
- Reviewed categorical columns.
- Verified data types.

No additional cleaning was required.

---

## Result

Dataset 3 was already well structured. The primary cleaning task was improving readability by replacing abbreviated categorical values with descriptive labels while preserving the original meaning of the data.

In [7]:
df_2["work_setting"].head(20)

0        Hybrid
1     In-person
2     In-person
3     In-person
4     In-person
5        Remote
6        Remote
7     In-person
8     In-person
9        Remote
10       Remote
11    In-person
12    In-person
13    In-person
14    In-person
15    In-person
16    In-person
17    In-person
18    In-person
19    In-person
Name: work_setting, dtype: object

In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", None)
print("Libraries imported successfully!")

#Get the data
df_1=pd.read_csv("../data/raw/DataAnalyst_set1_raw.csv")
df_2=pd.read_csv("../data/raw/jobs_in_data_set2_raw.csv")
df_3=pd.read_csv("../data/raw/global_ai_ml_data_salaries_set3_raw.csv")

#Start cleaning the first dataset
df_1.drop(columns=["Unnamed: 0"], inplace=True)
df_1.drop(columns=["Job Description"],inplace=True)

#changing the values of Easy Apply to 0's and 1's so its easyer to use in mysql
df_1["Easy Apply"]=df_1["Easy Apply"].replace({"True": 1 , "-1": 0})

#cleaning Salary Estimate and convirting it to three colums
df_1["Salary Estimate"] = df_1["Salary Estimate"].str.replace("(Glassdoor est.)","")
df_1["Salary Estimate"]=df_1["Salary Estimate"].str.replace("$","")
df_1["Salary Estimate"] =df_1["Salary Estimate"] .str.replace("K","")

#creat salary_min and salary_max out of Salary Estimate
df_1[["salary_min","salary_max"]] = df_1["Salary Estimate"].str.split("-",expand=True) 

#Deleting the row with the missing value so I could change the data type of both salary_min and salary_max
bad_row = df_1[df_1["Salary Estimate"] == "-1"].index
df_1=df_1.drop(bad_row)

#changing the data type
df_1["salary_min"]=df_1["salary_min"].astype(int)
df_1["salary_max"]=df_1["salary_max"].astype(int)

#Creat salary_avg
df_1["salary_avg"]=(df_1["salary_min"]+df_1["salary_max"])/2
print("------------------------------------------")
#print(df_1.columns)
# change the place of the colums
col1=df_1.pop("salary_min")
col2=df_1.pop("salary_max")
col3=df_1.pop("salary_avg")
df_1.insert(2,"salary_min",col1)
df_1.insert(3,"salary_max",col2)
df_1.insert(4,"salary_avg",col3)
#print(df_1.columns)

df_1.to_csv("../data/cleaned/data_analyst_set1_cleaned.csv",index=False)

#Data-set-2

#Converted company_size codes (S/M/L) to Small/Medium/Large.
df_2["company_size"] = df_2["company_size"].replace("S","Small")
df_2["company_size"] = df_2["company_size"].replace("M","Medium")
df_2["company_size"] = df_2["company_size"].replace("L","Large")

df_2.to_csv("../data/cleaned/jobs_in_data_set2_cleaned.csv",index=False)

#Data-set-3

#Expanded experience level abbreviations.
df_3["experience_level"] = df_3["experience_level"].replace("SE","Senior level")
df_3["experience_level"] = df_3["experience_level"].replace("MI","Mid level")
df_3["experience_level"] = df_3["experience_level"].replace("EN","Entry level")
df_3["experience_level"] = df_3["experience_level"].replace("EX","Executive level")

#Expanded employment type abbreviations.
df_3["employment_type"] = df_3["employment_type"].replace("FT","Full time")
df_3["employment_type"] = df_3["employment_type"].replace("PT","Part-time")
df_3["employment_type"] = df_3["employment_type"].replace("CT","Contract")
df_3["employment_type"] = df_3["employment_type"].replace("FL","Freelance")

#Expanded company size abbreviations.
df_3["company_size"] = df_3["company_size"].replace("L","Large")
df_3["company_size"] = df_3["company_size"].replace("M","Medium")
df_3["company_size"] = df_3["company_size"].replace("S","Small")

df_3.to_csv("../data/cleaned/global_ai_ml_data_salaries_set3_cleaned.csv",index=False)

#preparation to combine the three data sets

set1 = df_1.copy()
set2 = df_2.copy()
set3 = df_3.copy()

set1["company_size_standardized"] = set1["Size"].replace({
    "1 to 50 employees": "Small",
    "51 to 200 employees": "Medium",
    "201 to 500 employees": "Large",
    "501 to 1000 employees": "Large",
    "1001 to 5000 employees": "Large",
    "5001 to 10000 employees": "Large",
    "10000+ employees": "Large",
    "-1": "Unknown"
})

set1_standardized = pd.DataFrame({
    "work_year" : pd.NA,
    "job_title" : set1["Job Title"],
    "job_category" : pd.NA,
    "salary_currency" : "USA",
    "salary" : set1["salary_avg"]*1000,
    "salary_in_usd" : set1["salary_avg"]*1000,
    "employee_residence" : pd.NA,
    "experience_level" : pd.NA,
    "employment_type" : pd.NA,
    "remote_ratio" : pd.NA,
    "company_location" : set1["Location"],
    "company_size" : set1["company_size_standardized"],
    "data_source" : "Glassdoor Data Analyst Jobs"
})

set2["remote_ratio"] = set2["work_setting"].map({
    "In-person": 0,
    "Hybrid": 50,
    "Remote": 100
})

set2_standardized = pd.DataFrame({
    "work_year" : set2["work_year"],
    "job_title" : set2["job_title"],
    "job_category" : set2["job_category"],
    "salary_currency" : set2["salary_currency"],
    "salary" : set2["salary"],
    "salary_in_usd" : set2["salary_in_usd"],
    "employee_residence" : set2["employee_residence"],
    "experience_level" : set2["experience_level"],
    "employment_type" : set2["employment_type"],
    "work_setting" : set2["work_setting"],
    "company_location" : set2["company_location"],
    "company_size" : set2["company_size"],
    "data_source": "Jobs in Data"
})

set3["experience_level"] = set3["experience_level"].replace({
    "Entry level": "Entry-level",
    "Mid level": "Mid-level",
    "Senior level": "Senior",
    "Executive level": "Executive"
})

set3["employment_type"] = set3["employment_type"].replace({
    "Full time": "Full-time",
    "Part-time": "Part-time",
    "Contract": "Contract",
    "Freelance": "Freelance"
})

set3_standardized = pd.DataFrame({
    "work_year": set3["work_year"],
    "job_title": set3["job_title"],
    "job_category": pd.NA,
    "salary_currency": set3["salary_currency"],
    "salary": set3["salary"],
    "salary_in_usd": set3["salary_in_usd"],
    "employee_residence": set3["employee_residence"],
    "experience_level": set3["experience_level"],
    "employment_type": set3["employment_type"],
    "remote_ratio": set3["remote_ratio"],
    "company_location": set3["company_location"],
    "company_size": set3["company_size"],
    "data_source": "Global AI/ML Salaries"
})


combined_salary_data = pd.concat([
    set1_standardized,
    set2_standardized,
    set3_standardized

], ignore_index=True)
combined_salary_data = pd.concat([
    set1_standardized,
    set2_standardized,
    set3_standardized

], ignore_index=True)

#combined_salary_data["job_title"].combine_first(combined_salary_data["job_title"])

combined_salary_data.rename(
    columns={"salary": "salary_local"},
    inplace=True
)

combined_salary_data["work_setting_standardized"] = combined_salary_data["remote_ratio"].replace({
    0: "In-person",
    50: "Hybrid",
    100: "Remote"
})

combined_salary_data["work_setting_standardized"]= (
    combined_salary_data["work_setting_standardized"].
    combine_first(combined_salary_data["work_setting"]))

combined_salary_data.drop(columns=["work_setting"], inplace=True)

#print(combined_salary_data.shape)
print("----------------------------")
#print(combined_salary_data.isnull().sum())
print("----------------------------")
#print(combined_salary_data["data_source"].value_counts())
print("---------------------")
#print(combined_salary_data["salary_in_usd"].describe())
print("-----------------------------")
#print(combined_salary_data.columns.tolist())
print("------------------------------------------")
print(combined_salary_data.dtypes)

Libraries imported successfully!


C:\Users\abaho\AppData\Local\Temp\ipykernel_24332\1715515657.py:18: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_1["Easy Apply"]=df_1["Easy Apply"].replace({"True": 1 , "-1": 0})


------------------------------------------
----------------------------
----------------------------
---------------------
-----------------------------
------------------------------------------
work_year                     object
job_title                     object
job_category                  object
salary_currency               object
salary_local                 float64
salary_in_usd                float64
employee_residence            object
experience_level              object
employment_type               object
remote_ratio                 float64
company_location              object
company_size                  object
data_source                   object
work_setting_standardized     object
dtype: object


C:\Users\abaho\AppData\Local\Temp\ipykernel_24332\1715515657.py:167: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  combined_salary_data = pd.concat([
C:\Users\abaho\AppData\Local\Temp\ipykernel_24332\1715515657.py:173: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  combined_salary_data = pd.concat([


In [12]:
print(combined_salary_data.head())
print(combined_salary_data.columns)

  work_year                                          job_title job_category  \
0      <NA>  Data Analyst, Center on Immigration and Justic...         <NA>   
1      <NA>                               Quality Data Analyst         <NA>   
2      <NA>  Senior Data Analyst, Insights & Analytics Team...         <NA>   
3      <NA>                                       Data Analyst         <NA>   
4      <NA>                             Reporting Data Analyst         <NA>   

  salary_currency  salary_local  salary_in_usd employee_residence  \
0             USA       51500.0        51500.0                NaN   
1             USA       51500.0        51500.0                NaN   
2             USA       51500.0        51500.0                NaN   
3             USA       51500.0        51500.0                NaN   
4             USA       51500.0        51500.0                NaN   

  experience_level employment_type  remote_ratio company_location  \
0              NaN             NaN       